[🐛 Bug]: ConnectionFailedException: JdkWebSocket initial request execution error

### What happened?

An attempt to upgrade Selenium hub and Chrome nodes from version 4.25.0-20240922 to the latest ones shows that there is an error related to CDP - ConnectionFailedException: JdkWebSocket initial request execution error.
The issue is reproducible starting from 4.26.0-20241101 till the latest versions: 4.33.0-20250525.
The only one workaround to avoid the issue - set Selenium flag --enable-cdp false for Chrome nodes and disable all tests that use CDP features.


### Command used to start Selenium Grid with Docker (or Kubernetes)

```shell
version: "3"
services:
  selenium-hub:
    image: selenium/hub:4.32.0-20250505
    container_name: selenium-hub
    ports:
      - "4444:4444"
      - "4443:4443"
      - "4442:4442"
    environment:
      - SE_ENABLE_TRACING=false
    healthcheck:
      test: "/opt/bin/check-grid.sh --host 0.0.0.0 --port 4444"
      interval: "15s"
      timeout: "30s"
      retries: 5
    restart: always

  chrome:
    image: selenium/node-chrome:4.32.0-20250505
    shm_size: 2gb
    depends_on:
      - selenium-hub
    environment:
      - SE_EVENT_BUS_HOST=selenium-hub
      - SE_EVENT_BUS_PUBLISH_PORT=4442
      - SE_EVENT_BUS_SUBSCRIBE_PORT=4443
      - SE_VNC_NO_PASSWORD=1
      - SE_SCREEN_WIDTH=1600
      - SE_SCREEN_HEIGHT=900
      - SE_NODE_MAX_SESSIONS=1
      - SE_NODE_MAX_INSTANCES=1
      - SE_NODE_GRID_URL=http://${host}:4444
      - SE_OPTS=--enable-managed-downloads true --enable-cdp true
      - SE_ENABLE_TRACING=false
      - JAVA_OPTS=-Djava.net.preferIPv4Stack=true
    restart: always
    deploy:
      mode: replicated
      replicas: 8
```

### Relevant log output

```shell
Caused by: org.openqa.selenium.remote.http.ConnectionFailedException: JdkWebSocket initial request execution error
Build info: version: '4.32.0', revision: 'd17c8aa950'
System info: os.name: 'Linux', os.arch: 'amd64', os.version: '6.8.0-1024-aws', java.version: '17.0.5'
Driver info: driver.version: unknown
	at org.openqa.selenium.remote.http.jdk.JdkHttpClient.openSocket(JdkHttpClient.java:253)
	at org.openqa.selenium.devtools.Connection.<init>(Connection.java:89)
	at org.openqa.selenium.devtools.SeleniumCdpConnection.<init>(SeleniumCdpConnection.java:36)
	at org.openqa.selenium.devtools.SeleniumCdpConnection.lambda$create$2(SeleniumCdpConnection.java:103)
	at java.base/java.util.Optional.map(Optional.java:260)
	at org.openqa.selenium.devtools.SeleniumCdpConnection.create(SeleniumCdpConnection.java:103)
	at org.openqa.selenium.devtools.SeleniumCdpConnection.create(SeleniumCdpConnection.java:49)
	at org.openqa.selenium.devtools.DevToolsProvider.getImplementation(DevToolsProvider.java:50)
	at org.openqa.selenium.devtools.DevToolsProvider.getImplementation(DevToolsProvider.java:29)
	at org.openqa.selenium.remote.Augmenter.augment(Augmenter.java:207)
	at org.openqa.selenium.remote.Augmenter.augment(Augmenter.java:178)
... 13 more
Caused by: java.net.http.WebSocketHandshakeException
	at java.net.http/jdk.internal.net.http.websocket.OpeningHandshake.resultFrom(OpeningHandshake.java:226)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.postFire(CompletableFuture.java:614)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:844)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:483)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
Caused by: jdk.internal.net.http.websocket.CheckFailedException: Unexpected HTTP response status code 400
	at java.net.http/jdk.internal.net.http.websocket.OpeningHandshake.checkFailed(OpeningHandshake.java:343)
	at java.net.http/jdk.internal.net.http.websocket.OpeningHandshake.handleResponse(OpeningHandshake.java:252)
	at java.net.http/jdk.internal.net.http.websocket.OpeningHandshake.resultFrom(OpeningHandshake.java:222)
	... 10 more
```

### Operating System

Windows 11, Ubuntu 22.04.5 LTS, MacOS 15.5

### Docker Selenium version (image tag)

4.32.0-20250505

### Selenium Grid chart version (chart version)

-

Hi @VietND96,
I've tried several options to reproduce the problem in a new project with Selenium dependency only but with no effect.
However the issue with JDK initial request execution error is still reproducible on my production code that I can't share.
The exact configuration of a machine that executes tests: Selenium 4.32.0 (revision d17c8aa950), Ubuntu 22.04.5 LTS, JDK 17.0.14.
What I see in logs:

**Chrome node:**
`10:46:26.364 INFO [LocalNode.newSession] - Session created by the Node. Id: df8955314a44f1345a62f66ce0b58cd7, Caps: Capabilities {acceptInsecureCerts: true, browserName: chrome, browserVersion: 136.0.7103.59, chrome: {chromedriverVersion: 136.0.7103.49 (031848bc6ad0..., userDataDir: /tmp/.org.chromium.Chromium...}, fedcm:accounts: true, goog:chromeOptions: {debuggerAddress: localhost:41419}, networkConnectionEnabled: false, pageLoadStrategy: normal, platformName: linux, proxy: Proxy(), se:cdp: ws://localhost:4444/session..., se:cdpVersion: 136.0.7103.59, se:containerName: 2c81b6982de5, se:downloadsEnabled: true, se:gridWebSocketUrl: ws://localhost:3727/session..., se:noVncPort: 7900, se:vnc: ws://localhost:4444/session..., se:vncEnabled: true, se:vncLocalAddress: ws://172.18.0.3:7900, setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: accept, webSocketUrl: ws://localhost:4444/session..., webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:extension:minPinLength: true, webauthn:extension:prf: true, webauthn:virtualAuthenticators: true}
10:46:45.746 INFO [ProxyNodeWebsockets.createWsEndPoint] - Establishing connection to ws://localhost:3727/session/df8955314a44f1345a62f66ce0b58cd7
10:46:45.845 INFO [ProxyNodeWebsockets.createWsEndPoint] - Establishing connection to ws://localhost:41419/devtools/browser/a52f10f3-9ff2-4f44-9e9c-fa1f18ecc6ac
10:46:46.461 INFO [ProxyNodeWebsockets.createWsEndPoint] - Establishing connection to ws://localhost:3727/session/df8955314a44f1345a62f66ce0b58cd7
10:46:46.480 INFO [ProxyNodeWebsockets.createWsEndPoint] - Establishing connection to ws://localhost:41419/devtools/browser/a52f10f3-9ff2-4f44-9e9c-fa1f18ecc6ac
10:46:46.804 INFO [ProxyNodeWebsockets.createWsEndPoint] - Establishing connection to ws://localhost:41419/devtools/browser/a52f10f3-9ff2-4f44-9e9c-fa1f18ecc6ac
10:46:46.821 INFO [ProxyNodeWebsockets.createWsEndPoint] - Establishing connection to ws://localhost:3727/session/df8955314a44f1345a62f66ce0b58cd7
10:46:47.148 INFO [ProxyNodeWebsockets.createWsEndPoint] - Establishing connection to ws://localhost:41419/devtools/browser/a52f10f3-9ff2-4f44-9e9c-fa1f18ecc6ac
10:46:47.161 INFO [ProxyNodeWebsockets.createWsEndPoint] - Establishing connection to ws://localhost:3727/session/df8955314a44f1345a62f66ce0b58cd7
10:46:47.498 INFO [ProxyNodeWebsockets.createWsEndPoint] - Establishing connection to ws://localhost:41419/devtools/browser/a52f10f3-9ff2-4f44-9e9c-fa1f18ecc6ac
10:46:47.511 INFO [ProxyNodeWebsockets.createWsEndPoint] - Establishing connection to ws://localhost:3727/session/df8955314a44f1345a62f66ce0b58cd7
10:46:47.829 WARN [ProxyNodeWebsockets.apply] - Too many websocket connections initiated by df8955314a44f1345a62f66ce0b58cd7
10:46:48.073 WARN [ProxyNodeWebsockets.apply] - Too many websocket connections initiated by df8955314a44f1345a62f66ce0b58cd7`
**Selenium:**
` Caps: Capabilities {acceptInsecureCerts: true, browserName: chrome, browserVersion: 136.0.7103.59, chrome: {chromedriverVersion: 136.0.7103.49 (031848bc6ad0..., userDataDir: /tmp/.org.chromium.Chromium...}, fedcm:accounts: true, goog:chromeOptions: {debuggerAddress: localhost:44571}, networkConnectionEnabled: false, pageLoadStrategy: normal, platformName: linux, proxy: {}, se:cdp: ws://localhost:4444/session..., se:cdpVersion: 136.0.7103.59, se:containerName: 20d4d8a1b2fc, se:downloadsEnabled: true, se:gridWebSocketUrl: ws://localhost:18642/sessio..., se:noVncPort: 7900, se:vnc: ws://localhost:4444/session..., se:vncEnabled: true, se:vncLocalAddress: ws://172.18.0.9:7900, setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: accept, webSocketUrl: ws://localhost:4444/session..., webauthn:extension:credBlob: true, webauthn:extension:largeBlob: true, webauthn:extension:minPinLength: true, webauthn:extension:prf: true, webauthn:virtualAuthenticators: true}
10:46:47.831 WARN [ProxyWebsocketsIntoGrid.apply] - Connecting to upstream websocket failed
org.openqa.selenium.remote.http.ConnectionFailedException: JdkWebSocket initial request execution error
Build info: version: '4.32.0', revision: 'd17c8aa950'
System info: os.name: 'Linux', os.arch: 'amd64', os.version: '6.8.0-1024-aws', java.version: '21.0.6'
Driver info: driver.version: unknown
	at org.openqa.selenium.remote.http.jdk.JdkHttpClient.openSocket(JdkHttpClient.java:253)
	at org.openqa.selenium.remote.tracing.TracedHttpClient.openSocket(TracedHttpClient.java:45)
	at org.openqa.selenium.grid.router.ProxyWebsocketsIntoGrid.apply(ProxyWebsocketsIntoGrid.java:73)
	at org.openqa.selenium.grid.router.ProxyWebsocketsIntoGrid.apply(ProxyWebsocketsIntoGrid.java:43)
	at org.openqa.selenium.netty.server.WebSocketUpgradeHandler.handleHttpRequest(WebSocketUpgradeHandler.java:137)
	at org.openqa.selenium.netty.server.WebSocketUpgradeHandler.channelRead(WebSocketUpgradeHandler.java:100)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
	at io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.onHttpRequestChannelRead(WebSocketServerExtensionHandler.java:158)
	at io.netty.handler.codec.http.websocketx.extensions.WebSocketServerExtensionHandler.channelRead(WebSocketServerExtensionHandler.java:82)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:93)
	at io.netty.handler.codec.http.HttpServerKeepAliveHandler.channelRead(HttpServerKeepAliveHandler.java:64)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1357)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:868)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:796)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:732)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:658)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.net.http.WebSocketHandshakeException
	at java.net.http/jdk.internal.net.http.websocket.OpeningHandshake.resultFrom(OpeningHandshake.java:224)
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.postFire(CompletableFuture.java:614)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:844)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:483)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
Caused by: jdk.internal.net.http.websocket.CheckFailedException: Unexpected HTTP response status code 400
	at java.net.http/jdk.internal.net.http.websocket.OpeningHandshake.checkFailed(OpeningHandshake.java:341)
	at java.net.http/jdk.internal.net.http.websocket.OpeningHandshake.handleResponse(OpeningHandshake.java:250)
	at java.net.http/jdk.internal.net.http.websocket.OpeningHandshake.resultFrom(OpeningHandshake.java:220)`

The issue doesn't reproduce with Selenium 4.25. Started from 4.26 it is present.
May I ask you for an advice please? Maybe something related to websockets or CDP has been changes significatly? Or maybe you have any ideas how to debug this issue by myself (add more logs etc...) to understand why the Selenium complains on JDKWebSocket?

--- a/ENV_VARIABLES.md
+++ b/ENV_VARIABLES.md
@@ -144,4 +144,6 @@
 | SE_NODE_REGISTER_SHUTDOWN_ON_FAILURE | true | If this flag is enabled, the Node will shut down after the register period is completed. This is useful for container environments to restart and register again. If restarted multiple times, the Node container status will be CrashLoopBackOff | --register-shutdown-on-failure |
 | SE_NODE_RELAY_BROWSER_VERSION |  |  |  |
 | SE_NODE_RELAY_ONLY | true |  |  |
-| SE_EXTRA_LIBS | | Extra jars to add to the classpath | --ext |
+| SE_EXTRA_LIBS |  | Extra jars to add to the classpath in server component bootstrap | --ext |
+| SE_NODE_CONNECTION_LIMIT_PER_SESSION | 10 |  |  |
+| SE_SUPERVISORD_UNIX_SERVER_PASSWORD | secret |  |  |

--- a/NodeBase/Dockerfile
+++ b/NodeBase/Dockerfile
@@ -49,6 +49,7 @@ ENV LANG_WHICH=${LANG_WHICH} \
     SE_NODE_REGISTER_PERIOD="120" \
     SE_NODE_REGISTER_CYCLE="10" \
     SE_NODE_REGISTER_SHUTDOWN_ON_FAILURE="true" \
+    SE_NODE_CONNECTION_LIMIT_PER_SESSION="10" \
     SE_OTEL_SERVICE_NAME="selenium-node" \
     SE_NODE_RELAY_ONLY="true" \
     # Setting Selenium Manager to work offline

--- a/README.md
+++ b/README.md
@@ -1901,5 +1901,14 @@ mkdir /tmp/videos
 chown 1200:1201 /tmp/videos
 ```
 
+### Websocket connections per session get exhausted
+
+> org.openqa.selenium.remote.http.ConnectionFailedException: JdkWebSocket initial request execution error`
+
+This was reported in [#2850](https://github.com/SeleniumHQ/docker-selenium/issues/2850).
+Actually, from Grid version v4.26.0+, in Node CLI option `--connection-limit-per-session` (`SE_NODE_CONNECTION_LIMIT_PER_SESSION` environment variable) is set to `10` by default. Let X be the maximum number of websocket connections per session.This will ensure one session is not able to exhaust the connection limit of the host. Websocket connection might come from enable CDP, BiDi.
+
+Your test scenario or test framework implementation might be creating more than `X` connections per session, which will lead to the error above. You can optimize your implementation to use less connections per session, or you can increase the limit by setting the environment variable `SE_NODE_CONNECTION_LIMIT_PER_SESSION` to a value higher than `10` to allow more connections per session.
+
 ## Stargazers over time
 [![Stargazers over time](https://starchart.cc/SeleniumHQ/docker-selenium.svg?variant=adaptive)](https://starchart.cc/SeleniumHQ/docker-selenium)

--- a/scripts/generate_list_env_vars/description.yaml
+++ b/scripts/generate_list_env_vars/description.yaml
@@ -440,3 +440,12 @@
 - name: SE_NODE_RELAY_ONLY
   description: ''
   cli: ''
+- name: SE_EXTRA_LIBS
+  description: Extra jars to add to the classpath in server component bootstrap
+  cli: --ext
+- name: SE_NODE_CONNECTION_LIMIT_PER_SESSION
+  description: ''
+  cli: ''
+- name: SE_SUPERVISORD_UNIX_SERVER_PASSWORD
+  description: ''
+  cli: ''

--- a/scripts/generate_list_env_vars/value.yaml
+++ b/scripts/generate_list_env_vars/value.yaml
@@ -36,6 +36,8 @@
   default: '4443'
 - name: SE_EXTERNAL_URL
   default: ''
+- name: SE_EXTRA_LIBS
+  default: ''
 - name: SE_FRAME_RATE
   default: '15'
 - name: SE_GRID_URL
@@ -78,6 +80,8 @@
   default: ''
 - name: SE_NODE_BROWSER_VERSION
   default: stable
+- name: SE_NODE_CONNECTION_LIMIT_PER_SESSION
+  default: '200'
 - name: SE_NODE_CONTAINER_NAME
   default: ''
 - name: SE_NODE_DOCKER_CONFIG_FILENAME
@@ -240,6 +244,8 @@
   default: /tmp/supervisord.pid
 - name: SE_SUPERVISORD_START_RETRIES
   default: '5'
+- name: SE_SUPERVISORD_UNIX_SERVER_PASSWORD
+  default: secret
 - name: SE_UPLOAD_COMMAND
   default: ''
 - name: SE_UPLOAD_DESTINATION_PREFIX

